### PR TITLE
P4of round2

### PR DESCRIPTION
**Pull request by pierce-m**

### CODE REVIEW

**Comment by antoninbas** [comment]
Is this PR up for review?

**Comment by pierce-m** [comment]
Yes, trying to get openflow stuff working on bmv2

**Comment by pierce-m** [comment]
Can we please disable the style check for the of.py file? All it does is
initialize a dictionary

**Comment by antoninbas** [comment]
It is way easier to just fix the 2 lines with an issue

**Comment by pierce-m** [comment]
I'll try but in the past it hasn't been that simple

**Comment by antoninbas** [comment]
You should be able to install and run `flake8` on the file locally to test for style without having to go through the Travis regression every time.

File: p4c_bm/gen_pd.py [hunk 1]
```diff
@@ -305,6 +305,18 @@ def render_all_files(render_dict, gen_dir):
         with open(target, "w") as f:
             render_template(f, template, render_dict, _TEMPLATES_DIR,
                             prefix=_TENJIN_PREFIX)
+    if len(plugin_list) > 0:
+        for s in plugin_list:
+            plugin_dir = _PLUGIN_BASE_DIR + s
+            plugin_files = gen_file_lists(plugin_dir,
+                                          gen_dir+'/plugin/'+s)
```
Comment (antoninbas) [inline p4c_bm/gen_pd.py]: use `os.path.join` for `plugin_dir` and when calling `gen_file_lists`

**Comment by antoninbas** [comment]
Please add a test to https://github.com/p4lang/p4c-bm/blob/master/tests/test_p4c_bm.py which exercises your Python code (all of it), i.e. generates the OF plugin C code.

File: tests/test_p4c_bm.py [hunk 2]
```diff
@@ -143,6 +143,48 @@ def test_gen_pd(input_p4, tmpdir):
     assert set(expected_inc) == set([f.basename for f in inc_path.listdir()])
     assert set(expected_src) == set([f.basename for f in src_path.listdir()])
 
+def test_gen_of_pd(tmpdir):
+    cwd = os.getcwd()
```
Comment (antoninbas) [inline tests/test_p4c_bm.py]: there is no need for this, look at the other tests

File: tests/test_p4c_bm.py [hunk 3]
```diff
@@ -144,6 +144,49 @@ def test_gen_pd(input_p4, tmpdir):
...
+                        default=[],
+                        help="list of plugins to generate templates")
+    parser.add_argument('--openflow-mapping-dir',
+                        help="Directory of openflow mapping files")
+    parser.add_argument('--openflow-mapping-mod',
+                        help="Openflow mapping module name -- not a file name")
+    parser.plugin_list = ["of"]
+    parser.openflow_mapping_dir = os.path.join("tests", "of_mapping")
+    parser.openflow_mapping_mod = "l2_openflow"
+
```
Comment (antoninbas) [inline tests/test_p4c_bm.py]: Could you simply use the following:

```
from argparse import Namespace
args = Namespace(plugin_list=["of"], openflow_mapping_dir=os.path.join("tests", "of_mapping"),...)
```

thanks

**Comment by antoninbas** [comment]
don't forget to squash when merging

**Comment by pierce-m** [comment]
Is it ready?

File: tests/test_p4c_bm.py [hunk 4]
```diff
@@ -202,7 +237,6 @@ def test_main(tmpdir):
 
     # preprocessor flag
     assert call_main([input_p4, "-DANTONIN_ON"]) == 0
-    assert call_main([input_p4, "-I/home/"]) == 0
```
Comment (antoninbas) [inline tests/test_p4c_bm.py:205]: why remove?

**Comment by antoninbas** [comment]
actually I took one last look and you removed something when merging master into your branch, so make sure you undo that, I marked it in the diff

**Comment by antoninbas** [comment]
when this is done, the PR looks good to me
I still need to take one last look at the p4-build one, but this PR can be merged before the p4-build one, without breaking anything. Just make sure that you squash, or you can ask me to merge myself.

**Comment by pierce-m** [comment]
I don't know where that diff came from, I didn't remove that. It appeared
when I updated the branch via github

**Pull request merged**